pkg/webutil/hotwired.go

```diff
@@ -0,0 +1,126 @@
+package webutil
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"html/template"
+	"net/http"
+	"strings"
+
+	"github.com/go-redis/redis/v8"
+	"github.com/julienschmidt/httprouter"
+	"github.com/pkg/errors"
+	"github.com/rebuy-de/rebuy-go-sdk/v4/pkg/redisutil"
+	"github.com/sirupsen/logrus"
+	"nhooyr.io/websocket"
+)
+
+type Renderer interface {
+	Render(filename string, d interface{}) (*bytes.Buffer, error)
+}
+
+type HotwiredBroadcast[T any] struct {
+	broadcast *redisutil.Broadcast[HotwiredFrame[T]]
+	renderer  Renderer
+}
+
+func NewHotwiredBroadcast[T any](client redisutil.BroadcastRediser, key string, renderer Renderer) (*HotwiredBroadcast[T], error) {
+	redisBroadcast, err := redisutil.NewBroadcast[HotwiredFrame[T]](client, key)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	return &HotwiredBroadcast[T]{
+		broadcast: redisBroadcast,
+		renderer:  renderer,
+	}, nil
+}
+
+type HotwiredFrame[T any] struct {
+	Filename string
+	Value    *T
+}
+
+func (b *HotwiredBroadcast[T]) AddFrame(ctx context.Context, filename string, value *T) error {
+	err := b.broadcast.Add(ctx, &HotwiredFrame[T]{
+		Filename: filename,
+		Value:    value,
+	})
+	return errors.WithStack(err)
+}
+
+func (b *HotwiredBroadcast[T]) Handle(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	logrus.WithField("path", r.RequestURI).Debugf("accepting new connection")
+
+	c, err := websocket.Accept(w, r, &websocket.AcceptOptions{
+		InsecureSkipVerify: true,
+	})
+	if err != nil {
+		logrus.WithError(errors.WithStack(err)).Errorf("accepting websocket connection failed")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	defer c.Close(websocket.StatusInternalError, "fail")
+
+	id := "0-0"
+
+	for r.Context().Err() == nil {
+		v, newID, err := b.broadcast.Read(r.Context(), id)
+		if errors.Is(err, redis.Nil) {
+			continue
+		}
+		if err != nil {
+			logrus.WithError(err).Error(err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		id = newID
+
+		html, err := b.renderer.Render(v.Filename, v.Value)
+		if err != nil {
+			logrus.WithError(err).Error(err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		err = c.Write(r.Context(), websocket.MessageText, html.Bytes())
+		if err != nil {
+			logrus.WithError(err).Error(err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	c.Close(websocket.StatusNormalClosure, "")
+}
+
+func HotwiredTemplateFunctions() template.FuncMap {
+	return template.FuncMap{
+		"hotwiredImport": HotwiredImportTemplateFunction,
+		"hotwiredStream": HotwiredStreamTemplateFunction,
+	}
+}
+
+func HotwiredImportTemplateFunction() template.HTML {
+	return template.HTML(`
+      <script type="module">
+        import hotwiredTurbo from 'https://cdn.skypack.dev/@hotwired/turbo';
```

Comment:
I'm not a big fan of CDN hosted JS & CSS, however I'm not sure how else we should handle that.

Reply:
> 'm not a big fan of CDN hosted JS & CSS

Why exactly?

Reply:
Privacy, Security, depending on the CDN... also it's against the law, [at least for customer facing stuff](https://www.golem.de/news/landgericht-muenchen-einbindung-von-google-fonts-ist-rechtswidrig-2202-162826.html), imho all in all bad practice:
https://httptoolkit.tech/blog/public-cdn-risks/
https://shkspr.mobi/blog/2020/10/please-stop-using-cdns-for-external-javascript-libraries/

Reply:
We would require our own CDN then. Also our tools are not customer facing.

Reply:
You don't require your own CDN if you just put it in the source. But of course this would make it larger.

Reply:
I would not mind putting that to s3/cf. Or ship it directly. Not great, but not terrible.

Reply:
> just put it in the source

I'd like not to have a 22.6 kB text file in SCM (for Hotwired alone, not counting other deps like Bootstrap).

> I would not mind putting that to s3/cf. Or ship it directly. Not great, but not terrible.

Would work for me, but would require some access protection in case someone else uses this library.

Reply:
I will try Git LFS, but in another PR.